In [124]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import numpy as np
import os
import re
from scipy import stats as st
from statsmodels.stats.proportion import proportions_ztest as ztest

In [45]:
plt.style.use('fivethirtyeight')
pd.set_option('display.float_format', '{:.2f}'.format)
pd.set_option('display.min_rows', 20)
pd.set_option('display.max_columns', None)
pd.set_option('mode.chained_assignment', None)

# Test case #1

Assuming all files are identical in their structure being the segments of a larger chunk, we should be just fine with a loop concatenating the small portions back into one dataset. At four million rows we're unlikely to lack memory, however in such a case it's fairly easy to build in some data type optimization for each file.  
To make it easier we can put all the csv files into one folder created specifically for this purpose, and afterwards save the single file in case we need it later.

In [17]:
df = pd.DataFrame()

for file in os.scandir('csvs/'):
    print(f'Reading file: {file}')
    tmp = pd.read_csv(file)
    df = pd.concat([df, tmp])
    
df.info()

df.to_csv('the_whole_dataset.csv')

Reading file: <DirEntry 'test_csv - Copy (2).csv'>
Reading file: <DirEntry 'test_csv - Copy (3).csv'>
Reading file: <DirEntry 'test_csv - Copy (4).csv'>
Reading file: <DirEntry 'test_csv - Copy (5).csv'>
Reading file: <DirEntry 'test_csv - Copy (6).csv'>
Reading file: <DirEntry 'test_csv - Copy (7).csv'>
Reading file: <DirEntry 'test_csv - Copy (8).csv'>
Reading file: <DirEntry 'test_csv - Copy (9).csv'>
Reading file: <DirEntry 'test_csv - Copy.csv'>
Reading file: <DirEntry 'test_csv.csv'>
<class 'pandas.core.frame.DataFrame'>
Int64Index: 5761020 entries, 0 to 576101
Data columns (total 6 columns):
 #   Column             Dtype 
---  ------             ----- 
 0   HomeFname          object
 1   HomeLname          object
 2   Account            int64 
 3   HomeUserID         int64 
 4   HomeEventDatetime  object
 5   EventType          object
dtypes: int64(2), object(4)
memory usage: 307.7+ MB


# Test Case #2

## Results, before you read the rest

The first stab at this research was based on the assumption that past members contain the same valuable indicators as the ten users who did pay for a membership or a ticket.  

I compared the scarce numerical data and quantified several data points to look for measurable differences between the groups. The bottom line is I found next to nothing, and even the noticable difference in the number of subscribers which makes sense on paper, turns out to not be statistically significant.  

If you're still curious about my train of thought feel free to continue reading.

## First look

In [3]:
df = pd.read_excel('linkedin.xlsx')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 348 entries, 0 to 347
Data columns (total 74 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   linkedinProfile               348 non-null    object        
 1   Result                        348 non-null    object        
 2   Other                         191 non-null    object        
 3   email                         216 non-null    object        
 4   description                   278 non-null    object        
 5   headline                      348 non-null    object        
 6   location                      348 non-null    object        
 7   firstName                     348 non-null    object        
 8   lastName                      348 non-null    object        
 9   fullName                      348 non-null    object        
 10  subscribers                   347 non-null    float64       
 11  userId                        28

In [4]:
df.head()

linkedinProfile                   Result  \
0    https://www.linkedin.com/in/mark-kohoot-96a08b1  5. Purchased Membership   
1            https://www.linkedin.com/in/jonbowerman  5. Purchased Membership   
2           https://www.linkedin.com/in/parkeralynch  5. Purchased Membership   
3          https://www.linkedin.com/in/shacharschiff  5. Purchased Membership   
4  https://www.linkedin.com/in/gary-lacourt-811b1874        4.Purchased Event   

    Other                              email  \
0  Hunter                   mk@aeroscena.com   
1  Hunter                   jon@kidsdraw.org   
2  Hunter          Parker@HedgeHogHealth.com   
3  Hunter             shachar@badtesting.com   
4  Hunter  gary.lacourt@forevercompanies.com   

                                         description  \
0  Mark founded Aeroscena in 2010 with the missio...   
1  A creative problem solver who can rapidly desi...   
2                                                NaN   
3  I'm the founder and principal consultant at Ba...   
4  Serial entrepreneur and business founder. Main...   

                                            headline  \
0          Bio-pharmaceutical / functional fragrance   
1  Co Founder and CTO - Product Development and M...   
2  CEO at HedgeHog Health | 2020-2021 Global Educ...   
3  I believe that high-quality, powerful software...   
4                 Founder and CEO, Forever Companies   

                              location firstName  lastName        fullName  \
0       Cleveland, Ohio, United States      Mark    Kohoot     Mark Kohoot   
1  Farmington, Michigan, United States       Jon  Bowerman    Jon Bowerman   
2   Royal Oak, Michigan, United States    Parker     Lynch    Parker Lynch   
3          Dallas-Fort Worth Metroplex   Shachar    Schiff  Shachar Schiff   
4  Milwaukee, Wisconsin, United States      Gary   LaCourt    Gary LaCourt   

   subscribers       userId  \
0      1704.00   3172258.00   
1      1745.00  14910006.00   
2      2549.00  22364061.00   
3      1696.00  35484653.00   
4       710.00 265539469.00   

                               mutualConnectionsText  \
0                 1 mutual connection: Eric Sullivan   
1                 1 mutual connection: Eric Sullivan   
2                                                NaN   
3  2 mutual connections: Andrew Storz and Eric Su...   
4                                                NaN   

                 mailFromDropcontact  \
0                   mk@aeroscena.com   
1                   jon@kidsdraw.org   
2          parker@hedgehoghealth.com   
3             shachar@badtesting.com   
4  gary.lacourt@forevercompanies.com   

                                      company  \
0                                   Aeroscena   
1                               Kids Draw Bob   
2                             HedgeHog Health   
3  BadTesting® — Better Software. On Time.   
4                           Forever Companies   

                                          companyUrl  \
0        https://www.linkedin.com/company/aeroscena/   
1      https://www.linkedin.com/company/kidsdrawbob/   
2   https://www.linkedin.com/company/hedgehoghealth/   
3       https://www.linkedin.com/company/badtesting/   
4  https://www.linkedin.com/company/forever-compa...   

                    jobTitle  \
0                        CEO   
1  President and Illustrator   
2    Chief Executive Officer   
3      Founder and Principal   
4              Founder & CEO   

                                      jobDescription  \
0  Aeroscena develops and markets fragrances from...   
1  The energy in a room of children who've found ...   
2  Revolutionizing education technology and welln...   
3  At BadTesting, we exist because of a simple be...   
4               Founder and CEO of Forever Companies   

                    jobLocation          jobDateRange  \
0    Cleveland/Akron, Ohio Area  Sep 2010 – Present   
1          Bloomfield Hills, MI  Jul 2019 – Present   
2                   

In [5]:
df.sample(5)

linkedinProfile                Result  \
156            https://www.linkedin.com/in/alexpurtell  2. Sent Personal f/u   
337  https://www.linkedin.com/in/levi-baker-a54610109/           2021 Member   
29          https://www.linkedin.com/in/johannesariens           3. Approved   
281            https://www.linkedin.com/in/divatommei/           2021 Member   
222  https://www.linkedin.com/in/sean-holmes-703a4538/           2021 Member   

      Other                      email  \
156  Hunter    alex@horizontwolabs.com   
337     NaN                        NaN   
29   Hollis           ja@routeline.com   
281     NaN  diva.tommei@eitdigital.eu   
222     NaN  sean@nectarsunglasses.com   

                                           description  \
156                                                NaN   
337                                                NaN   
29   If anything I’m up to sounds interesting or ...   
281                                                NaN   
222  Experienced Co-Founder with a demonstrated his...   

                                      headline  \
156  Partner & Growth Lead at Horizon Two Labs   
337           Builder at Creative elements Inc   
29       CEO at Route Line - Create & Execute    
281            Head of EIT Digital Accelerator   
222             Co-Founder - Nectar Sunglasses   

                                      location firstName lastName  \
156              Columbus, Ohio, United States      Alex  Purtell   
337            Heber City, Utah, United States      Levi    Baker   
29                        Greater Seattle Area  Johannes   Ariens   
281                        Rome, Latium, Italy      Diva   Tommei   
222  Charleston, South Carolina, United States      Sean   Holmes   

            fullName  subscribers       userId mutualConnectionsText  \
156     Alex Purtell      3304.00 189407345.00                   NaN   
337       Levi Baker        34.00 458370784.00                   NaN   
29   Johannes Ariens      1239.00 137417395.00                   NaN   
281      Diva Tommei      5601.00  73384298.00                   NaN   
222      Sean Holmes      2466.00          NaN                   NaN   

           mailFromDropcontact                company  \
156    alex@horizontwolabs.com       Horizon Two Labs   
337                        NaN  Creative elements Inc   
29            ja@routeline.com             Route Line   
281                        NaN            EIT Digital   
222  sean@nectarsunglasses.com      Nectar Sunglasses   

                                           companyUrl  \
156  https://www.linkedin.com/company/horizontwolabs/   
337                                               NaN   
29       https://www.linkedin.com/company/route-line/   
281     https://www.linkedin.com/company/eit-digital/   
222        https://www.linkedin.com/company/33296600/   

                            jobTitle  \
156   Partner + Accelerator Director   
337                          Builder   
29                               CEO   
281  Head of EIT Digital Accelerator   
222     Co-founder Nectar Sunglasses   

                                        jobDescription  \
156  At Horizon Two Labs we define the adjacent pos...   
337                                                NaN   
29   Route Line is an end-to-end adventure company ...   
281                                                NaN   
222                                                NaN   

                            jobLocation                        jobDateRange  \
156                 Columbus, Ohio Area                Aug 2018 – Present   
337                                 NaN                                 NaN   
29   Seattle, Washington, United States                Jul 2021 – Present   
281                 Rome, Latium, Italy                Jan 2022 – Present   
222                                 NaN  Oct 2011 - Present · 10 yrs 4 mos   

              company2                                  

In [6]:
df.duplicated().sum()

0

We have 348 LinkedIn profiles, with no duplicates at least at first glance. A lot of missing data with no way of filling in the gaps, so we'll take a closer look on how it's distributed in our target groups, as well as look for potential categorical data points.

## Examining data

In [7]:
df.Result.value_counts()

2021 Member                157
2. Sent Personal f/u       127
3. Approved                 28
Out of Sequence             26
4.Purchased Event            6
5. Purchased Membership      4
Name: Result, dtype: int64

So if we count both ideal and second best outcomes our conversion rate is ~3% which sounds normal for an outreach, however it leaves us with very little data; so we'll have to examine past members too as they have at some point paid for a membership - why they are no longer doing so is another question (assuming they have not indeed renewed their membership).

In [12]:
won = df[(df['Result']=='5. Purchased Membership') | (df['Result']=='4.Purchased Event')]
won

linkedinProfile                   Result  \
0    https://www.linkedin.com/in/mark-kohoot-96a08b1  5. Purchased Membership   
1            https://www.linkedin.com/in/jonbowerman  5. Purchased Membership   
2           https://www.linkedin.com/in/parkeralynch  5. Purchased Membership   
3          https://www.linkedin.com/in/shacharschiff  5. Purchased Membership   
4  https://www.linkedin.com/in/gary-lacourt-811b1874        4.Purchased Event   
5  https://www.linkedin.com/in/lyle-stoflet-38a07ab8        4.Purchased Event   
6          https://www.linkedin.com/in/whiteandrewdc        4.Purchased Event   
7           https://www.linkedin.com/in/donniemcfall        4.Purchased Event   
8                https://www.linkedin.com/in/sykraft        4.Purchased Event   
9      https://www.linkedin.com/in/alan-dahl-2a6781b        4.Purchased Event   

    Other                              email  \
0  Hunter                   mk@aeroscena.com   
1  Hunter                   jon@kidsdraw.org   
2  Hunter          Parker@HedgeHogHealth.com   
3  Hunter             shachar@badtesting.com   
4  Hunter  gary.lacourt@forevercompanies.com   
5  Hunter      stoflet@stratusindustries.com   
6  Hunter                                NaN   
7  Hunter                                NaN   
8  Hollis                     sy@sykraft.com   
9  Hollis                                NaN   

                                         description  \
0  Mark founded Aeroscena in 2010 with the missio...   
1  A creative problem solver who can rapidly desi...   
2                                                NaN   
3  I'm the founder and principal consultant at Ba...   
4  Serial entrepreneur and business founder. Main...   
5  What’s to love about our design products?\n\...   
6  Helping families and individuals reach their G...   
7                                                NaN   
8  I untie knots and as a brand building professi...   
9                                                NaN   

                                            headline  \
0          Bio-pharmaceutical / functional fragrance   
1  Co Founder and CTO - Product Development and M...   
2  CEO at HedgeHog Health | 2020-2021 Global Educ...   
3  I believe that high-quality, powerful software...   
4                 Founder and CEO, Forever Companies   
5  Managing Partner at Stratus Industries / Gear ...   
6                             Founder at Align & Co.   
7  Owner at McFall Consulting, Co-Founder at The ...   
8                        Founder of Fantastic Brands   
9                        Director, Advisor, Investor   

                                location firstName     lastName  \
0         Cleveland, Ohio, United States      Mark       Kohoot   
1    Farmington, Michigan, United States       Jon     Bowerman   
2     Royal Oak, Michigan, United States    Parker        Lynch   
3            Dallas-Fort Worth Metroplex   Shachar       Schiff   
4    Milwaukee, Wisconsin, United States      Gary      LaCourt   
5    Milwaukee, Wisconsin, United States      Lyle      Stoflet   
6       Holland, Michigan, United States    Andrew  White, D.C.   
7  Minneapolis, Minnesota, United States    Donnie       McFall   
8          Los Angeles Metropolitan Area        Sy        Kraft   
9              Atlanta Metropolitan Area      Alan         Dahl   

             fullName  subscribers       userId  \
0         Mark Kohoot      1704.00   3172258.00   
1        Jon Bowerman      1745.00  14910006.00   
2        Parker Lynch      2549.00  22364061.00   
3      Shachar Schiff      1696.00  35484653.00   
4        Gary LaCourt       710.00 265539469.00   
5        Lyle Stoflet       343.00 418200730.00   
6  Andrew White, D.C.       799.00 167264628.00   
7       Donnie McFall      1842.00  21326084.00   
8            Sy Kraft     12731.00  14176975.00   
9           Alan Dahl       928.00  34755678.00   

                               mutualConnectionsText  \
0                 1 mutual

In [13]:
churned = df[df['Result']=='2021 Member']

In [19]:
churned.sample(5)

linkedinProfile       Result Other  \
306             https://www.linkedin.com/in/jayfaires/  2021 Member   NaN   
265          https://www.linkedin.com/in/chriskresser/  2021 Member   NaN   
288  https://www.linkedin.com/in/eric-rieger-a559a010/  2021 Member   NaN   
320        https://www.linkedin.com/in/joshualevittnd/  2021 Member   NaN   
194         https://www.linkedin.com/in/maxwellhertan/  2021 Member   NaN   

                          email  \
306                         NaN   
265  chris@kresserinstitute.com   
288                         NaN   
320                         NaN   
194                         NaN   

                                           description  \
306  Faires created TWA (The Wellness Agency), whic...   
265  Chris Kresser, M.S., L.Ac, is a practitioner o...   
288  CRNA - In partnership with gastroenterologist ...   
320  Naturopathic physician and medical director at...   
194  Lover of entrepreneurship, self-help, philosop...   

                                              headline  \
306                          The Wellness Agency (TWA)   
265  Co-Founder and President at California Center ...   
288  Sales & Market Dev @ Atrantil - Ambassador of ...   
320                             Naturopathic Physician   
194  Founder @ Silvi & Megaphone Marketing (AFR Fas...   

                                          location firstName lastName  \
306  Los Angeles County, California, United States       Jay   Faires   
265             Oakland, California, United States     Chris  Kresser   
288                    Dallas-Fort Worth Metroplex      Eric   Rieger   
320             Hamden, Connecticut, United States    Joshua   Levitt   
194        Santa Monica, California, United States   Maxwell   Hertan   

           fullName  subscribers      userId mutualConnectionsText  \
306      Jay Faires      2564.00 16437770.00                   NaN   
265   Chris Kresser      2963.00 68898932.00                   NaN   
288     Eric Rieger       393.00 38280161.00                   NaN   
320   Joshua Levitt       356.00 36833015.00                   NaN   
194  Maxwell Hertan      7243.00         NaN                   NaN   

    mailFromDropcontact              company  \
306                 NaN  The Wellness Agency   
265                 NaN    Kresser Institute   
288                 NaN             Atrantil   
320                 NaN         Whole Health   
194                 NaN            My Silvi    

                                            companyUrl  \
306  https://www.linkedin.com/company/the-wellness-...   
265                                                NaN   
288         https://www.linkedin.com/company/atrantil/   
320                                                NaN   
194         https://www.linkedin.com/company/71044201/   

                                     jobTitle  \
306                                   Founder   
265                           Founder and CEO   
288           PR-Sales and Market Development   
320  Naturopathic Physician, Medical Director   
194                                Co-Founder   

                                        jobDescription  \
306  The Wellness Agency (TWA) empowers wellness bu...   
265                                                NaN   
288  Developed by gastroenterologist Ken Brown, MD ...   
320  Natural soultions for common and complex medic...   
194                                                NaN   

                jobLocation                      jobDateRange  \
306              Venice, Ca              Jan 2018 – Present   
265  Las Vegas, Nevada Area              Oct 2015 – Present   
288              Dallas, TX              Feb 2015 – Present   
320                     NaN              Mar 2006 – Present   
194                     NaN  Aug 2020 - Present · 1 yr 6 mos   

                                      company2  \
306                                  Lionsgate   
265  California Center for Functio

Seeing a lot of: "head honchos" (founder/owner/C-level etc.), also first-name email addresses. Subscribers, twitter, education and skills seem to vary a lot, as well as age and location. All seem to be within US, suggesting this was a US event.  
Let's take a look at the rest and see if any of that is different.

In [20]:
lost = df[(df.index.isin(won.index)==False) & (df.index.isin(churned.index)==False)]

In [30]:
lost.sample(5)

linkedinProfile                Result  \
11      https://www.linkedin.com/in/anne-zizzo-0075603           3. Approved   
15   https://www.linkedin.com/in/samantha-friedman-...           3. Approved   
129              https://www.linkedin.com/in/schelkidd  2. Sent Personal f/u   
87   https://www.linkedin.com/in/gary-peters-568585140  2. Sent Personal f/u   
30     https://www.linkedin.com/in/todd-heath-351a1611           3. Approved   

      Other email                                        description  \
11   Hunter   NaN  At Zizzo Group, we believe that successful eng...   
15   Hunter   NaN                                                NaN   
129  Hunter   NaN  Two words can describe Larschelby; Passion & E...   
87   Hunter   NaN  I am a professional writer and former newspape...   
30   Hollis   NaN  My goal is to lead a positive and productive l...   

                                              headline  \
11   Founder & CEO at Zizzo Group - Engagement Mark...   
15                                  Manager, Marketing   
129  Independent Boisset Collection Ambassador at B...   
87                                Founder, Final Draft   
30                 Publisher + Art Director, Bomb Snow   

                                           location         firstName  \
11              Milwaukee, Wisconsin, United States              Anne   
15              Birmingham, Michigan, United States          Samantha   
129                Kohler, Wisconsin, United States  Larschelby Kidd    
87   Sheboygan Falls Town, Wisconsin, United States              Gary   
30              Great Falls, Montana, United States              Todd   

              lastName                            fullName  subscribers  \
11               Zizzo                          Anne Zizzo      2775.00   
15            Friedman                   Samantha Friedman      1176.00   
129   WisconsinWineGuy  Larschelby Kidd   WisconsinWineGuy       772.00   
87              Peters                         Gary Peters       487.00   
30               Heath                          Todd Heath      1516.00   

          userId mutualConnectionsText mailFromDropcontact  \
11   10326535.00                   NaN                 NaN   
15   72697048.00                   NaN                 NaN   
129  83909867.00                   NaN                 NaN   
87  574728416.00                   NaN                 NaN   
30   41337709.00                   NaN                 NaN   

                      company  \
11        Zizzo Ventures, LLC   
15            The Mom Project   
129        Boisset Collection   
87                Final Draft   
30   Freelance, Self-Employed   

                                            companyUrl  \
11                                                 NaN   
15   https://www.linkedin.com/company/the-mom-project/   
129  https://www.linkedin.com/company/boisset-colle...   
87                                                 NaN   
30                                                 NaN   

                                      jobTitle  \
11                                     Founder   
15             Seller Marketing Manager @ Etsy   
129  Independent Boisset Collection Ambassador   
87                        Founder, Final Draft   
30        Art Director, Designer, Pixel Pusher   

                                        jobDescription  \
11                    Investment, leasing, development   
15                                                 NaN   
129  I am a Independent Boisset Collection Ambassad...   
87                                                 NaN   
30   I specialize in layout and catalog design for ...   

                             jobLocation          jobDateRange  \
11   Milwaukee, Wisconsin, United States  May 2021 – Present   
15     Brooklyn, New York, United States  Jul 2021 – Present   
129                            Wisconsin  Jan 2019 – Present   
87                                   NaN  Apr 2017 

A quick look confirms the impression of the outreach being targeted at business leaders, superficially there is little to no difference between the profiles. Next we'll pick a few characteristicts to compare won, churned and lost members/attendees to see if we can detect any patterns.

In [102]:
good_combined = pd.concat([won,churned])

In [103]:
dfs = (('won',won), ('churned',churned), ('combined positive', good_combined), ('lost',lost)) #for mass processing

## Detecting patterns

In [104]:
for frame in dfs:
    print(f'Examining {frame[0]} leads.')
    print(f'Mean subscribers: {frame[1].subscribers.mean()}, median subscribers: {frame[1].subscribers.median()}')
    print(f'Mean user id: {frame[1].userId.mean()}')
    print('='*100)

Examining won leads.
Mean subscribers: 2504.7, median subscribers: 1700.0
Mean user id: 99719454.2
Examining churned leads.
Mean subscribers: 2621.121794871795, median subscribers: 1470.5
Mean user id: 114916101.24489796
Examining combined positive leads.
Mean subscribers: 2614.10843373494, median subscribers: 1525.0
Mean user id: 113509004.2962963
Examining lost leads.
Mean subscribers: 2628.1215469613257, median subscribers: 1175.0
Mean user id: 152985565.2872928


We have little numerical data - user ids are barely that, although they might suggest that more seasoned LinkedIn users (= smaller id) tend to be more receptive to the events and membership.  
While mean number of subscribers is almost identical across the three groups, medians vary significantly - with larger numbers indicating a propensity towards membership or at least attending an event.  
Next we'll quantify what we can and mine for possible categorical data.

In [105]:
for frame in dfs:   
    frame[1]['desc_len'] = frame[1].description.apply(lambda x: 0 if x is np.nan else len(x))
    frame[1]['headline_len'] = frame[1].headline.apply(lambda x: 0 if x is np.nan else len(x))
    print(f'Examining {frame[0]} leads.')
    print(f'Mean length of headline: {frame[1].headline_len.mean()}, median length of headline: {frame[1].headline_len.median()}')
    print(f'Mean length of description: {frame[1].desc_len.mean()}, median length of description: {frame[1].desc_len.median()}')

Examining won leads.
Mean length of headline: 62.0, median length of headline: 50.0
Mean length of description: 570.7, median length of description: 398.5
Examining churned leads.
Mean length of headline: 65.59235668789809, median length of headline: 49.0
Mean length of description: 756.6751592356688, median length of description: 432.0
Examining combined positive leads.
Mean length of headline: 65.37724550898204, median length of headline: 49.0
Mean length of description: 745.5389221556886, median length of description: 432.0
Examining lost leads.
Mean length of headline: 59.005524861878456, median length of headline: 50.0
Mean length of description: 684.5911602209944, median length of description: 440.0


The length of sections themselves shows little variabiltiy, although it's curious that won and previously won leads show a slight tendency to write longer headlines and shorter descriptions. Next we'll look at contents of descriptions and headlines to see if there's any difference in branding oneself that correlates with purchasing membership or an event ticket.

In [106]:
def inspect_titles(text):
    titles = (('owner','owner'),
        ('founder|co-founder','founder'),
        ('ceo| chief executive','ceo'),
        ('c[a-z]{2}| chief .+ officer','c-level'),
        ('entrepren','entrepreneur'),
        ('managing [a-z]+| director| s*vp| head of ','higher_management'))
    
    title_list = []
    for pair in titles:
        if text is not np.nan:
            if re.search(pair[0], text.lower()):
                title_list.append(pair[1])
    return title_list

In [107]:
for frame in dfs:
    frame[1]['title_list_desc']= frame[1].description.apply(inspect_titles)
    frame[1]['title_list_headline']= frame[1].headline.apply(inspect_titles)
    frame[1]['desc_titles'] = frame[1].title_list_desc.apply(lambda x: len(x))
    frame[1]['headline_titles'] = frame[1].title_list_headline.apply(lambda x: len(x))
    print(f'Examining {frame[0]} leads.')
    print(f'Mean number of titles in headline: {frame[1].headline_titles.mean()}, median number of titles in headline: {frame[1].headline_titles.median()}')
    print(f'Mean number of titles in description: {frame[1].desc_titles.mean()}, median number of titles in description: {frame[1].desc_titles.median()}')

Examining won leads.
Mean number of titles in headline: 1.7, median number of titles in headline: 1.5
Mean number of titles in description: 1.1, median number of titles in description: 1.0
Examining churned leads.
Mean number of titles in headline: 1.5987261146496816, median number of titles in headline: 1.0
Mean number of titles in description: 1.5286624203821657, median number of titles in description: 1.0
Examining combined positive leads.
Mean number of titles in headline: 1.6047904191616766, median number of titles in headline: 1.0
Mean number of titles in description: 1.5029940119760479, median number of titles in description: 1.0
Examining lost leads.
Mean number of titles in headline: 1.4364640883977902, median number of titles in headline: 1.0
Mean number of titles in description: 1.4751381215469612, median number of titles in description: 1.0


In [108]:
for col in df.columns:
    for frame in dfs:
        print(f'Leads: {frame[0]}, missing values in {col}: {round(((frame[1][col].isna().sum())/frame[1].shape[0]),2)*100}%')
    print('='*100)

Leads: won, missing values in linkedinProfile: 0.0%
Leads: churned, missing values in linkedinProfile: 0.0%
Leads: combined positive, missing values in linkedinProfile: 0.0%
Leads: lost, missing values in linkedinProfile: 0.0%
Leads: won, missing values in Result: 0.0%
Leads: churned, missing values in Result: 0.0%
Leads: combined positive, missing values in Result: 0.0%
Leads: lost, missing values in Result: 0.0%
Leads: won, missing values in Other: 0.0%
Leads: churned, missing values in Other: 100.0%
Leads: combined positive, missing values in Other: 94.0%
Leads: lost, missing values in Other: 0.0%
Leads: won, missing values in email: 30.0%
Leads: churned, missing values in email: 36.0%
Leads: combined positive, missing values in email: 35.0%
Leads: lost, missing values in email: 40.0%
Leads: won, missing values in description: 30.0%
Leads: churned, missing values in description: 25.0%
Leads: combined positive, missing values in description: 26.0%
Leads: lost, missing values in descr

Doesn't look like gaps in any of the aspects of profiles can tell us anything meaningful about the difference between the groups.

In [109]:
for frame in dfs:
    frame[1]['jobDateRange2'] = frame[1]['jobDateRange2'].fillna(' ')
    frame[1]['two_jobs'] = frame[1].jobDateRange2.apply(lambda x: True if 'Present' in str(x) else False)
    print(f'Examining {frame[0]} leads.')
    print(f'Share of people combining jobs: {round((frame[1].two_jobs.sum()/frame[1].shape[0]),2)}')

Examining won leads.
Share of people combining jobs: 0.9
Examining churned leads.
Share of people combining jobs: 0.6
Examining combined positive leads.
Share of people combining jobs: 0.62
Examining lost leads.
Share of people combining jobs: 0.7


It appears that having two jobs at once used to somewhat reduce the chance of becoming a member; however it's not yet clear how significant this correlation is.

## Statistical significance

A couple of differences we were able to spot warrant running test of statistical significance on. We will be comparing two groups of profiles: those who purchased membership or a ticket event + those who were members in the past, and those who failed to do either of those after the last outreach.  
We'll begin with the number of subscribers - we suspect that people cultivating a larger number of followers on LinkedIn are more likely to become members and attend events as the prime networking and self-promoting opportunities.  
We'll start with a formal check of normality of distribution using the Shapiro-Wilk test, H0 being that the data is distributed normally, and H1 being that the distribution is not normal.

In [113]:
alpha = 0.01
dfs_2 = (('combined positive', good_combined), ('lost',lost))

In [120]:
for frame in dfs_2:
    print(f'Checking distribution in {frame[0]}:')
    p_val = st.shapiro(frame[1].subscribers.dropna()).pvalue
    if  p_val > alpha:
        print(f'P-value: {p_val}. Can not reject the null hypothesis, the number of subscribers is distributed normally.')
    else:
        print(f'P-value: {p_val}. The null hypothesis should be rejected, the number of subscribers is distributed not normally.')

Checking distribution in combined positive:
P-value: 1.1725375345625936e-17. The null hypothesis should be rejected, the number of subscribers is distributed not normally.
Checking distribution in lost:
P-value: 1.0155986513698732e-22. The null hypothesis should be rejected, the number of subscribers is distributed not normally.


As the data is not distributed normally we will proceed with using the The Mann-Whitney U test to check whether the observed difference is significant.  
H0: the average number of subscribers in both groups is the same.  
H1: the average number of subscribers in the 'good_combined' group is higher.

In [123]:
x = good_combined.subscribers.dropna()
y = lost.subscribers.dropna()
p_val = st.mannwhitneyu(x,y, alternative='greater').pvalue
if  p_val < alpha:
    print(f'P-value: {p_val}. We reject the null hypothesis, the average number of subsribers in "good_combined" is higher.')
else:
    print(f'P-value: {p_val}. We can not reject the null hypothesis, there is no significant difference between the numbers of subscribers in two groups.')

P-value: 0.09707386827171183. We can not reject the null hypothesis, there is no significant difference between the numbers of subscribers in two groups.
